eras/conway/impl/src/Cardano/Ledger/Conway/Rules/GovCert.hs

```diff
@@ -236,57 +240,80 @@ conwayGovCertTransition = do
         isCurrentMember || isPotentialFutureMember ?! ConwayCommitteeIsUnknown coldCred
         pure
-          vState
-            { vsCommitteeState =
-                CommitteeState
-                  { csCommitteeCreds = Map.insert coldCred newMemberState csCommitteeCreds
+          certState
+            { certVState =
+                vState
+                  { vsCommitteeState =
+                      CommitteeState
+                        { csCommitteeCreds = Map.insert coldCred newMemberState csCommitteeCreds
+                        }
                   }
             }
   case cert of
     ConwayRegDRep cred deposit mAnchor -> do
       Map.notMember cred vsDReps ?! ConwayDRepAlreadyRegistered cred
       deposit == ppDRepDeposit ?! ConwayDRepIncorrectDeposit deposit ppDRepDeposit
+      let drepState =
+            DRepState
+              { drepExpiry =
+                  computeDRepExpiryVersioned
+                    cgcePParams
+                    cgceCurrentEpoch
+                    (vState ^. vsNumDormantEpochsL)
+              , drepAnchor = mAnchor
+              , drepDeposit = ppDRepDeposit
+              , drepDelegs = mempty
+              }
       pure
-        vState
-          { vsDReps =
-              Map.insert
-                cred
-                ( DRepState
-                    ( computeDRepExpiryVersioned
-                        cgcePParams
-                        cgceCurrentEpoch
-                        (vState ^. vsNumDormantEpochsL)
-                    )
-                    mAnchor
-                    ppDRepDeposit
-                )
-                vsDReps
+        certState
+          { certVState =
+              vState
+                { vsDReps = Map.insert cred drepState vsDReps
+                }
           }
     ConwayUnRegDRep cred refund -> do
-      case Map.lookup cred vsDReps of
-        Nothing -> failBecause $ ConwayDRepNotRegistered cred
-        Just drepState ->
-          let paidDeposit = drepState ^. drepDepositL
-           in refund == paidDeposit ?! ConwayDRepIncorrectRefund refund paidDeposit
-      pure vState {vsDReps = Map.delete cred vsDReps}
+      let mDRepState = Map.lookup cred vsDReps
+          drepRefundMismatch = do
+            drepState <- mDRepState
+            let paidDeposit = drepState ^. drepDepositL
+            guard (refund /= paidDeposit)
+            pure paidDeposit
+      isJust mDRepState ?! ConwayDRepNotRegistered cred
+      failOnJust drepRefundMismatch $ ConwayDRepIncorrectRefund refund
+      let
+        certState' =
+          certState {certVState = vState {vsDReps = Map.delete cred vsDReps}}
+      pure $
+        case mDRepState of
+          Nothing -> certState'
+          Just dRepState ->
+            certState'
+              { certDState =
+                  certDState
+                    { dsUnified = drepDelegs dRepState UM.⋪ UM.DRepUView (dsUnified certDState)
+                    }
+              }
```

Comment:
Maybe slightly easier to follow using lenses?   But I can see this module is not so "lensified" - it doesn't matter, just an idea. 
```suggestion
           certDStateL . dsUnifiedL %~  ((UM.⋪) (drepDelegs dRepState) . UM.DRepUView)
```